marklogic-client-api/src/main/java/com/marklogic/client/impl/HTTPSamlAuthInterceptor.java

```diff
@@ -29,96 +29,98 @@
-	    if(authorizer!=null) {
-	        if(expiringSAMLAuth==null) {
-	            authorize(null);
-	        } else if(threshold<=Instant.now().getEpochSecond()){
-	            authorize(expiringSAMLAuth.getExpiry());
-	                    
-	        }
-	    }
-	    else if(renewer!=null) {
-	        if(threshold<=Instant.now().getEpochSecond()){
-	            HTTPSamlAuthInterceptor httpSamlAuthInterceptor = new HTTPSamlAuthInterceptor(expiringSAMLAuth, renewer);
-	            if(isCallbackExecuting.compareAndSet(false, true))
-	                Executors.defaultThreadFactory().newThread(httpSamlAuthInterceptor).start();
-	            
-	        }
-	    }
-	    String samlHeaderValue = RESTServices.AUTHORIZATION_TYPE_SAML+ " "+ RESTServices.AUTHORIZATION_PARAM_TOKEN + "=" + authorizationTokenValue;
-	    Request authenticatedRequest = request.newBuilder().header(RESTServices.HEADER_AUTHORIZATION, samlHeaderValue).build();
-	    return chain.proceed(authenticatedRequest);
-	  }
-	  
-	  private synchronized void authorize(Instant expiry) {
-		  if(expiry == null && expiringSAMLAuth != null) {
-			  return;
-		  }
-		  if(expiry!=null && expiry!=expiringSAMLAuth.getExpiry()) {
-			  return;
-		  }
-		  checkAuthorizationExpiry(expiry);
-		  expiringSAMLAuth = authorizer.apply(expiringSAMLAuth);
-		  if(expiringSAMLAuth==null) {
-			  throw new IllegalArgumentException("SAML Authentication cannot be null");
-		  }
-		  
-		  if(expiringSAMLAuth.getAuthorizationToken() == null) {
-			  throw new IllegalArgumentException("SAML Authentication token cannot be null");
-		  }
-		  authorizationTokenValue = expiringSAMLAuth.getAuthorizationToken();
-		  
-		  
-		  if(isCallbackExecuting.compareAndSet(false, true))
-		      threshold = expiry.getEpochSecond();
-		  
-	  }
-	  
-	  private static void checkAuthorizationExpiry(Instant instant) {
-	      if(instant == null) {
-              throw new IllegalArgumentException("SAML authentication does not have expiry value.");
-          }
-          if(instant.isBefore(Instant.now())) {
-              throw new IllegalArgumentException("SAML authentication token has expired.");
-          }
-	  }
-	  static void renew(ExpiringSAMLAuth expiringSamlAuth) {
-	      try {
-	      Instant newInstant = renewer.apply(expiringSamlAuth);
-	      
-	      checkAuthorizationExpiry(newInstant);
-	      long current = Instant.now().getEpochSecond();
-	      threshold = current+ ((newInstant.getEpochSecond() - current)/2);
-	      } finally {
-            isCallbackExecuting = new AtomicBoolean(false);
-        }
-	  }
     @Override
-    public void run() {
-        renew(expiringSAMLAuth);
-        
+    public Response intercept(Chain chain) throws IOException {
+        Request request = chain.request();
+        if (authorizer != null) {
+            if ((expiringSAMLAuth == null || threshold <= Instant.now().getEpochSecond())
+                    && isCallbackExecuting.compareAndSet(false, true)) {
```

Comment:
Apologies, but it sank in that using compareAndSet() for the authorizer was a bad suggestion.

The problem is that we need to block all threads until the authorizer gets the new authorization token.

In other words, the previous synchronized approach for the authorizer was better.